main.cpp

```diff
@@ -70,28 +70,485 @@ write 3 UDTs below that EACH have:
+
+    void useToneAlgorithm();
+};
+
+ToneControl::ToneControl() 
+{
+    std::cout << "ToneControl being constructed" << std::endl;
+}
+
+ToneControl::ToneAlgorithm::ToneAlgorithm(): color1(10), color2(11), color3(12)
+{
+    std::cout << "ToneAlgorithm constructing with upper limit: " <<  upperLimit << "and lower limit: " << lowerLimit <<std::endl;
+}
+
+void ToneControl::ToneAlgorithm::setToneColors(int c1, int c2, int c3)
+{
+    std::cout << "tone color set to " << c1 << ", " << c2 << ", " << c3 << std::endl;
+}
+
+bool ToneControl::ToneAlgorithm::setUpperLimit(float upperLimit1)
+{
+    std::cout << upperLimit1 << std::endl;
+    return true;
+}
+
+bool ToneControl::ToneAlgorithm::setLowerLimit(float lowerLimit1)
+{
+    std::cout << lowerLimit1 << std::endl;
+    return true;
+}
+
+float ToneControl::setToneLevel(float input, float adjust)
+{
+    std::cout << input << " is good" << std::endl;
+    float toReturn = input - adjust;
+    std::cout << (input - adjust) << " is adjusted amount" << std::endl;
+
+    return toReturn;
+}
+
+bool ToneControl::requireRepair()
+{
+    return true;
+}
+
+bool ToneControl::autoAdjust()
+{
+    return true;
+}
+ToneControl::~ToneControl()
+{
+    std::cout <<"ToneControl destructor" << std::endl;
+}
+ToneControl::ToneAlgorithm::~ToneAlgorithm()
+{
+    std::cout << "ToneAlgorithm destructing" << std::endl;
+}
+
+void ToneControl::ToneAlgorithm::incrementAndSetColors(int color11, int color22, int color33)
+{
+    setToneColors(color11, color22, color33);
+    std::cout << "before a while" << toIncrement;
+    while(toIncrement < 10)
+    {
+        setToneColors(color11 + toIncrement, color22 + toIncrement, color33 + toIncrement);
+        //do nothing
+        std::cout << "doing nothing but this" << std::endl;
+        ++toIncrement;
+    }
+    std::cout << "after a while" << toIncrement;            
+}
+
+void ToneControl::changeToneAlgorithm()
+{
+    if(toneAlgoType == 'A')
+    {
+        toneAlgorithms.incrementAndSetColors(30,40,50);
+        toneAlgoType = 'B';
+    }
+    else
+    {
+        toneAlgoType = 'A';
+        toneAlgorithms.incrementAndSetColors(20,30,52);
+    }        
+}
+
+void ToneControl::useToneAlgorithm()
+{
+    changeToneAlgorithm();
+}
 
 /*
- new UDT 4:
+ new UDT 4: MusicMachine
  with 2 member functions
  */
+struct MusicMachine
+{
+    Computer computer = Computer();
+    ToneControl toneControl = ToneControl();
+
+    MusicMachine();
+    ~MusicMachine();
+
+    void powerOnComputer(Computer computer1);
+    void adjustToneControl(ToneControl toneControl1);
+    void powerOffComputer(Computer computer1);
+    void endComputing(Computer computer);
+};
+
+MusicMachine::MusicMachine()
+{
+    std::cout << "MusicMachine constructed" << std::endl;
+}
+MusicMachine::~MusicMachine()
+{
+    std::cout << "ending computing" << std::endl;
+    this->endComputing(this->computer);
+    std::cout << "MusicMachine destructed " << std::endl;
+}
```

Comment:
the `this` keyword has not been covered yet.
I do not want other students seeing `this->` used in project parts before Project 5 Part 2.
If this `this->` usage is left in, they will see it when you share your project to the Project 5 Part 1 channel.
- [ ] revise